internal/processor/stream/processor.go

```diff
@@ -212,31 +218,70 @@ func (p *Processor) readBatch(
 // Callers must not access result concurrently with HandleStream.
 func (p *Processor) HandleStream(
 	ctx context.Context,
+	async bool,
 	baseEvent model.APMEvent,
-	reader io.Reader,
+	reader io.ReadCloser,
 	batchSize int,
 	processor model.BatchProcessor,
 	result *Result,
 ) error {
-	// Since processor.ProcessBatch can block for some time until all the batch
-	// is added to the modelindexer's cache and there isn't a fail-fast rejection,
-	// we want to cap how many in-flight requests are read at any time.
+	// Limits the number of concurrent batch decodes.
 	// Defaults to 200 (N), only allowing N requests to read and cache Y events
-	// (determined by batchSize) from the batch, effectively limitting the decoding
-	// concurrency to N batches at any time, which should be a good ceiling. The
-	// ceiling also reduces the contention on the modelindexer.activeMu.
-	select {
-	case p.sem <- struct{}{}:
-	case <-ctx.Done():
-		return ctx.Err()
+	// (determined by batchSize) from the batch.
+	// The ceiling may also reduce the contention on the modelindexer.activeMu.
+	// Clients can set a async to true which makes the processor handle the
+	// events in the background. Returns with an error `publish.ErrFull` if the
+	// semaphore is full.
+	if err := p.acquireLock(ctx, async); err != nil {
```

Comment:
(I might be lacking context here) The use of semaphore seems to limit the concurrency which, in effect, also controls the load APM-Server is processing at any moment. An async operation will defeat this purpose (if it is intended).

*brain dump*: I think we can eliminate semaphore entirely and add a middleware that implements [token based rate limiter](https://pkg.go.dev/golang.org/x/time/rate) for achieving a similar end result as what is expected from the semaphore. We can additionally expose the configuration of a rate limiter so that users can configure it based on their specific workload. WDYT?

Reply:
> (I might be lacking context here) The use of semaphore seems to limit the concurrency which, in effect, also controls the load APM-Server is processing at any moment. An async operation will defeat this purpose (if it is intended).

The semaphore won't be released back until the goroutine has finished work when `async=true`. So the semaphore will still have the desired effect. The only difference is in case the server is under load, then the semaphore won't be acquired and `publish.ErrFull` is returned.

> brain dump: I think we can eliminate semaphore entirely and add a middleware(?) that implements [token based rate limiter](https://pkg.go.dev/golang.org/x/time/rate) for achieving a similar end result as what is expected from the semaphore. We can additionally expose the configuration of a rate limiter so that users can configure it based on their specific workload. WDYT?

I had some thoughts on moving the semaphore to a middleware, but after this feature I'm not so sure. It seems to add a fair bit of complexity.

I'm not sure the rate limiter fits entirely with the limitation that we're trying to achieve here. The rate limiter would limit on a period of time (which?), while the semaphore is agnostic to time and always limits the concurrent number of decodes in the server. 

Since the rate limiter would be limiting based on a time period, which we don't want, it wouldn't matter to it if the events are happening serially (very fast) or concurrently.